runtime-sdk/src/modules/core/mod.rs

```diff
@@ -129,51 +132,50 @@ const CONTEXT_KEY_GAS_USED: &str = "core.GasUsed";
 
 impl Module {
     /// Initialize state from genesis.
-    fn init(ctx: &mut DispatchContext<'_>, genesis: &Genesis) {
+    fn init<C: Context>(ctx: &mut C, genesis: &Genesis) {
         // Set genesis parameters.
         Self::set_params(ctx.runtime_state(), &genesis.parameters);
     }
 
     /// Migrate state from a previous version.
-    fn migrate(_ctx: &mut DispatchContext<'_>, _from: u32) -> bool {
+    fn migrate<C: Context>(_ctx: &mut C, _from: u32) -> bool {
         // No migrations currently supported.
         false
     }
 }
 
 impl API for Module {
-    fn use_gas<C: Context>(ctx: &mut C, gas: u64) -> Result<(), Error> {
-        let new_gas_used = match (
-            ctx.tx_auth_info().map(|ai| ai.fee.gas),
-            ctx.tx_value::<u64>(CONTEXT_KEY_GAS_USED).copied(),
-        ) {
-            (Some(gas_limit), Some(gas_used)) => {
-                let sum = gas_used.checked_add(gas).ok_or(Error::GasOverflow)?;
-                if sum > gas_limit {
-                    return Err(Error::OutOfGas);
-                }
-                Some(sum)
-            }
-            (None, None) => None,
-            _ => panic!("inconsistent tx availability"),
-        };
-
+    fn use_batch_gas<C: Context>(ctx: &mut C, gas: u64) -> Result<(), Error> {
         let batch_gas_limit = Self::params(ctx.runtime_state()).max_batch_gas;
-        let batch_gas_used = ctx.value::<u64>(CONTEXT_KEY_GAS_USED);
+        let batch_gas_used = ctx.value::<u64>(CONTEXT_KEY_GAS_USED).or_default();
         let batch_new_gas_used = batch_gas_used
             .checked_add(gas)
             .ok_or(Error::BatchGasOverflow)?;
         if batch_new_gas_used > batch_gas_limit {
+            println!("batch gas limit: {:?}", batch_gas_limit);
```

Comment:
do we have some logging facility more formal than this

Reply:
https://crates.io/crates/tracing is good

Reply:
Yes, the context has a `slog` logging facility, see the `get_logger` method.

Reply:
The `tracing` crate looks nice though, maybe we should use it instead of `slog`.